shared-module/displayio/__init__.c

```diff
@@ -19,14 +20,83 @@
 
 primary_display_t displays[CIRCUITPY_DISPLAY_LIMIT];
 
-static inline void swap(int16_t* a, int16_t* b) {
-    int16_t temp = *a;
-    *a = *b;
-    *b = temp;
+bool refresh_area(displayio_display_obj_t* display, const displayio_area_t* area) {
+    uint16_t buffer_size = 512;
+
+    displayio_area_t clipped;
+    // Clip the area to the display by overlapping the areas. If there is no overlap then we're done.
+    if (!displayio_display_clip_area(display, area, &clipped)) {
+        return true;
+    }
+    uint16_t subrectangles = 1;
+    uint16_t rows_per_buffer = displayio_area_height(&clipped);
+    if (displayio_area_size(area) > buffer_size) {
+        rows_per_buffer = buffer_size / displayio_area_width(&clipped);
+        subrectangles = displayio_area_height(&clipped) / rows_per_buffer;
+        if (displayio_area_height(&clipped) % rows_per_buffer != 0) {
+            subrectangles++;
+        }
+        buffer_size = rows_per_buffer * displayio_area_width(&clipped);
+    }
+    uint32_t buffer[buffer_size / 2];
+    uint16_t remaining_rows = displayio_area_height(&clipped);
+
+    for (uint16_t j = 0; j < subrectangles; j++) {
+        displayio_area_t subrectangle = {
+            .x1 = clipped.x1,
+            .y1 = clipped.y1 + rows_per_buffer * j,
+            .x2 = clipped.x2,
+            .y2 = clipped.y1 + rows_per_buffer * (j + 1)
+        };
+        if (remaining_rows < rows_per_buffer) {
+            subrectangle.y2 = subrectangle.y1 + remaining_rows;
+        }
+        remaining_rows -= rows_per_buffer;
+
+        displayio_display_begin_transaction(display);
+        displayio_display_set_region_to_update(display, subrectangle.x1, subrectangle.y1,
+                                                        subrectangle.x2, subrectangle.y2);
+        displayio_display_end_transaction(display);
+
+        uint32_t mask[(buffer_size / 32) + 1];
+        for (uint16_t k = 0; k < (buffer_size / 32) + 1; k++) {
```

Comment:
Would `memset()` be faster?

Reply:
maybe not - I looked on stackoverflow, and, it depends.

Reply:
easier to read either way. :-)

Reply:
Didn't work with naive memset. This has the advantage of knowing the array type and that it is aligned.